crates/driver/src/domain/competition/bad_tokens/metrics.rs

```diff
@@ -1,11 +1,67 @@
-use {super::Quality, crate::domain::eth};
+use {super::Quality, crate::domain::eth, dashmap::DashMap, std::sync::Arc};
 
+/// Monitors tokens to determine whether they are considered "unsupported" based
+/// on the ratio of failing to total settlement encoding attempts. A token must
+/// have participated in at least `REQUIRED_MEASUREMENTS` attempts to be
+/// evaluated. If, at that point, the ratio of failures is greater than or equal
+/// to `FAILURE_RATIO`, the token is considered unsupported.
 #[derive(Default)]
-pub struct Detector;
+pub struct Detector(Arc<Inner>);
+
+#[derive(Default)]
+struct Inner {
+    counter: DashMap<eth::TokenAddress, TokenStatistics>,
+}
+
+#[derive(Default)]
+struct TokenStatistics {
+    attempts: u32,
+    fails: u32,
+}
 
 impl Detector {
-    pub fn get_quality(&self, _token: eth::TokenAddress) -> Option<Quality> {
-        // TODO implement a reasonable heuristic
-        None
+    /// The ratio of failures to attempts that qualifies a token as unsupported.
+    const FAILURE_RATIO: f64 = 0.9;
+    /// The minimum number of attempts required before evaluating a token’s
+    /// quality.
+    const REQUIRED_MEASUREMENTS: u32 = 20;
+
+    pub fn get_quality(&self, token: eth::TokenAddress) -> Option<Quality> {
+        let measurements = self.0.counter.get(&token)?;
+        if measurements.attempts >= Self::REQUIRED_MEASUREMENTS
+            && (measurements.fails as f64 / measurements.attempts as f64) >= Self::FAILURE_RATIO
+        {
+            Some(Quality::Unsupported)
+        } else {
+            None
+        }
+    }
+
+    /// Updates the tokens that participated in settlements by
+    /// incrementing their attempt count.
+    /// `failure` indicates whether the settlement was successful or not.
+    pub fn update_tokens(
+        &self,
+        token_pairs: Vec<(eth::TokenAddress, eth::TokenAddress)>,
```

Comment:
I wanted to avoid redundant cloning when calling `Map::entry` function, which requires the owned type, and if that’s not sufficient, let’s adhere to the idiomatic approach that unnecessarily forces copying values.

Reply:
In current solution with `Vec<>` as function argument particular token is still copied in call to `Map::entry`. It is copied because ownership of vector item cannot be taken from `Vec` and moved to function or variable. This is because vector stores its data in continuous heap memory buffer and to pass any vector item by value to function requires copy of this item from heap to stack.
So changing to slice as argument will not introduce any additional copy.

Reply:
> particular token is still copied in call to `Map::entry`. It is moved, not copied.

Wait, what? Ownership of vector items can be moved from a Vec to a function or variable when the Vec is consumed (e.g., via `.into_iter()`), and moving items does not involve copying unless explicitly cloned. `Map::entry` works with owned values, so no copying occurs when items are moved directly.

Reply:
Yes, you are right, `into_iter()` can move all items by direct iterating over the vector data buffer. I had on my mind case of moving one item from the vector (using index operator for instance) which is not possible.

I was playing a bit with sample code to check how passing by value and reference looks after compilation to see in which case `memcpy` is used, and it looks that for code without optimizations function parameters are passed by registers (if they fit) or copied by `memcpy` (if size is too large to use with registers) and it is valid for both pass-by value and pass-by reference. Basically the code for both functions looks similar. After enabling release level optimizations there is no `memcpy` for smaller and larger data size and direct memory access is used for iteration. Code of pass-by value and reference looks also very similar.
I think above is valid if we are using types which implements Copy trait, in other case this probably looks different (for instance for String type).
If you want to look at the asm here is the code which I was using: https://godbolt.org/z/ejGbYjcP4

Summarizing: In my opinion we should stick with idiomatic approach and compiler will do the optimal final result without any memory copy in this case.